File: Templates/trie.cpp
#include <bits/stdc++.h>

using namespace std;


class Node{
	public:
	vector<Node*> children;
	bool isComplete;
	Node(){
		children.resize(26);
		isComplete = false;
	}

	bool containsKey(char key){
		return children[key - 'a'] != NULL;
	}

	void addKey(char key){
		children[key - 'a'] = new Node();
	}

	Node *getNode(char key){
		return children[key - 'a'];
	}



};

class Trie {
public:
    Node *root;
    Trie() {
        root = new Node();
    }

    void insert(string word) {
	int N = word.length();
	Node *dummy = root;
	for(int i = 0; i < N; i++){
		char key = word[i];
		if(!dummy->containsKey(key)){
			dummy->addKey(key);
		}
		dummy = dummy->getNode(key);
	}
	dummy->isComplete = true;
    }

    bool search(string word) {
	int N = word.length();
	Node *dummy = root;
	for(int i = 0; i < N; i++){
		char key = word[i];
		if(!dummy->containsKey(key)){
			return false;
		}
		dummy = dummy->getNode(key);
	}
	return dummy->isComplete;
    }

    bool startsWith(string prefix) {
	int N = prefix.length();
	Node *dummy = root;
	for(int i = 0; i < N; i++){
		char key = prefix[i];
		if(!dummy->containsKey(key)){
			return false;
		}
		dummy = dummy->getNode(key);
	}
	return true;
    }
};

File: Templates/Trie.py
class Node:
    def __init__(self):
        self.children = [None] * 26
        self.terminal = False

class Trie:
    def __init__(self):
        self.root = Node()

    def add_word(self, word: str) -> None:
        curr = self.root
        for i in range(len(word)):
            if not curr.children[ord(word[i]) - 97]:
                curr.children[ord(word[i]) - 97] = Node()
            curr = curr.children[ord(word[i]) - 97]
        curr.terminal = True

    def search(self, word: str) -> bool:
        curr = self.root
        for i in range(len(word)):
            if curr.children[ord(word[i]) - 97]:
                curr = curr.children[ord(word[i]) - 97]
            else:
                return False
        return curr.terminal

    def prefix_exist(self, prefix: str) -> bool:
        curr = self.root
        for i in range(len(prefix)):
            if curr.children[ord(prefix[i]) - 97]:
                curr = curr.children[ord(prefix[i]) - 97]
            else:
                return False
        return True
